File: api/match-service/src/app/services/match_service.py
from typing import List, Dict, Any, Optional
from sqlalchemy import text
from app.config.database import db
from app.models import User, UserInteraction, Match
from app.services.vector_matching_service import VectorMatchingService

class MatchService:
    """Service layer for matching operations"""
    
    def __init__(self):
        """Initialize match service"""
        self.vector_service = VectorMatchingService()
    
    def get_user_matches(self, user_id: int, limit: int = 10) -> List[Dict[str, Any]]:
        """
        Get basic potential matches for a user (fallback method)
        """
        """
        Get potential matches for a user
        
        Args:
            user_id: ID of the user to find matches for
            limit: Maximum number of matches to return
            
        Returns:
            List of potential matches
        """
        # Basic matching algorithm based on sexual preferences and location
        
        # Get target user's preferences
        target_user = User.query.filter_by(id=user_id).first()
        if not target_user:
            return []
        
        target_gender = target_user.gender
        target_pref = target_user.sex_pref
        
        # Build query for compatible users
        query = User.query.filter(
            User.id != user_id,
            User.latitude.isnot(None),
            User.longitude.isnot(None)
        )
        
        # Apply compatibility filters
        if target_pref == 'both':
            # Target likes everyone, find people who like target's gender
            query = query.filter(
                db.or_(
                    User.sex_pref == target_gender,
                    User.sex_pref == 'both'
                )
            )
        else:
            # Target likes specific gender, find people of that gender who like target's gender
            query = query.filter(
                User.gender == target_pref,
                db.or_(
                    User.sex_pref == target_gender,
                    User.sex_pref == 'both'
                )
            )
        
        # Order and limit results
        compatible_users = query.order_by(
            User.fame.desc(),
            User.created_at.desc()
        ).limit(limit).all()
        
        # Convert to dict format
        return [{
            'id': user.id,
            'username': user.username,
            'first_name': user.first_name,
            'age': user.age,
            'bio': user.bio,
            'fame': user.fame,
            'algorithm_type': 'basic'
        } for user in compatible_users]
    
    def like_user(self, user_id: int, target_user_id: int) -> Dict[str, Any]:
        """
        Like another user
        
        Args:
            user_id: ID of the user doing the liking
            target_user_id: ID of the user being liked
            
        Returns:
            Dictionary with like result and potential match status
        """
        return self.vector_service.record_interaction(user_id, target_user_id, 'like')
    
    def unlike_user(self, user_id: int, target_user_id: int) -> Dict[str, Any]:
        """
        Unlike another user
        
        Args:
            user_id: ID of the user doing the unliking
            target_user_id: ID of the user being unliked
            
        Returns:
            Dictionary with unlike result
        """
        return self.vector_service.record_interaction(user_id, target_user_id, 'pass')
    
    def block_user(self, user_id: int, target_user_id: int) -> Dict[str, Any]:
        """
        Block another user
        
        Args:
            user_id: ID of the user doing the blocking
            target_user_id: ID of the user being blocked
            
        Returns:
            Dictionary with block result
        """
        result = self.vector_service.record_interaction(user_id, target_user_id, 'block')
        
        # Deactivate any existing match
        match = Match.query.filter(
            db.or_(
                db.and_(Match.user1_id == user_id, Match.user2_id == target_user_id),
                db.and_(Match.user1_id == target_user_id, Match.user2_id == user_id)
            )
        ).first()
        
        if match:
            match.is_active = False
            db.session.commit()
            result['match_deactivated'] = True
        
        return result
    
    def run_matching_algorithm(self, user_id: int, algorithm_type: str = 'vector_based', 
                              limit: int = 20, max_distance: Optional[int] = None,
                              age_range: Optional[tuple] = None) -> List[Dict[str, Any]]:
        """
        Run sophisticated matching algorithm for a user
        
        Args:
            user_id: ID of the user to find matches for
            algorithm_type: Type of algorithm to use ('vector_based', 'basic')
            limit: Maximum number of matches to return
            max_distance: Maximum distance in km
            age_range: Tuple of (min_age, max_age)
            
        Returns:
            List of ranked matches with compatibility scores
        """
        if algorithm_type == 'vector_based':
            return self.vector_service.get_potential_matches(
                user_id, limit, max_distance, age_range
            )
        else:
            # Fallback to basic matching
            return self.get_user_matches(user_id, limit)

File: api/match-service/src/services/match_service.go
package services

// MatchResult represents the result of a matching operation with full profile data
type MatchResult struct {
	ID                 int      `json:"id"`
	Username           string   `json:"username"`
	FirstName          string   `json:"first_name"`
	Age                int      `json:"age"`
	Bio                string   `json:"bio"`
	Fame               int      `json:"fame"`
	AlgorithmType      string   `json:"algorithm_type"`
	CompatibilityScore *float64 `json:"compatibility_score,omitempty"`
	Distance           *float64 `json:"distance,omitempty"`
}

// MatchCandidate represents a simplified matching result with only essential data
type MatchCandidate struct {
	ID                 int      `json:"id"`
	AlgorithmType      string   `json:"algorithm_type"`
	CompatibilityScore *float64 `json:"compatibility_score,omitempty"`
	Distance           *float64 `json:"distance,omitempty"`
}

// AgeRange represents an age filter range
type AgeRange struct {
	Min int `json:"min"`
	Max int `json:"max"`
}

// MatchService provides high-level matching operations using specialized services
type MatchService struct {
	userService        *UserService
	interactionService *InteractionService
	algorithmService   *AlgorithmService
}

// NewMatchService creates a new MatchService with all dependencies
func NewMatchService() *MatchService {
	return &MatchService{
		userService:        NewUserService(),
		interactionService: NewInteractionService(),
		algorithmService:   NewAlgorithmService(),
	}
}

// GetUserMatches retrieves active matches for a user
func (s *MatchService) GetUserMatches(userID int) ([]MatchResult, error) {
	return s.userService.GetUserMatches(userID)
}

// LikeUser records a like interaction and handles match creation
func (s *MatchService) LikeUser(userID, targetUserID int) (map[string]interface{}, error) {
	return s.interactionService.LikeUser(userID, targetUserID)
}

// UnlikeUser removes a like interaction and deactivates matches
func (s *MatchService) UnlikeUser(userID, targetUserID int) (map[string]interface{}, error) {
	return s.interactionService.UnlikeUser(userID, targetUserID)
}

// BlockUser blocks a user and removes any existing matches
func (s *MatchService) BlockUser(userID, targetUserID int) (map[string]interface{}, error) {
	return s.interactionService.BlockUser(userID, targetUserID)
}

// RunMatchingAlgorithm executes the specified matching algorithm and returns full profile data
func (s *MatchService) RunMatchingAlgorithm(userID int, algorithmType string, limit int, maxDistance *int, ageRange *AgeRange) ([]MatchResult, error) {
	request := BuildMatchingRequest(userID, algorithmType, limit, maxDistance, ageRange)
	return s.algorithmService.RunMatchingAlgorithm(request)
}

// GetMatchingCandidates executes the specified matching algorithm and returns only user IDs with scores
func (s *MatchService) GetMatchingCandidates(userID int, algorithmType string, limit int, maxDistance *int, ageRange *AgeRange) ([]MatchCandidate, error) {
	request := BuildMatchingRequest(userID, algorithmType, limit, maxDistance, ageRange)
	return s.algorithmService.GetMatchingCandidates(request)
}

// GetUserInteractions retrieves all interactions for a user
func (s *MatchService) GetUserInteractions(userID int) ([]interface{}, error) {
	interactions, err := s.interactionService.GetUserInteractions(userID)
	if err != nil {
		return nil, err
	}
	
	// Convert to interface{} slice for JSON serialization compatibility
	result := make([]interface{}, len(interactions))
	for i, interaction := range interactions {
		result[i] = interaction
	}
	
	return result, nil
}

// GetInteractionBetweenUsers gets interaction between two specific users
func (s *MatchService) GetInteractionBetweenUsers(userID, targetUserID int) (map[string]interface{}, error) {
	interaction, err := s.interactionService.GetInteractionBetweenUsers(userID, targetUserID)
	if err != nil {
		return nil, err
	}
	
	return map[string]interface{}{
		"user_id":          interaction.UserID,
		"target_user_id":   interaction.TargetUserID,
		"interaction_type": interaction.InteractionType,
		"created_at":       interaction.CreatedAt,
	}, nil
}

// GetMutualLikes finds users who mutually liked each other
func (s *MatchService) GetMutualLikes(userID int) ([]int, error) {
	return s.interactionService.GetMutualInteractions(userID)
}

// IsUserBlocked checks if a user is blocked by another user
func (s *MatchService) IsUserBlocked(userID, targetUserID int) (bool, error) {
	return s.interactionService.IsUserBlocked(userID, targetUserID)
}

// GetAvailableAlgorithms returns all available matching algorithms
func (s *MatchService) GetAvailableAlgorithms() []AlgorithmInfo {
	return s.algorithmService.GetAvailableAlgorithms()
}

// GetAlgorithmPerformanceStats returns performance statistics for algorithms
func (s *MatchService) GetAlgorithmPerformanceStats() map[string]interface{} {
	return s.algorithmService.GetAlgorithmPerformanceStats()
}

// InvalidateUserCaches clears all cached data for a user
func (s *MatchService) InvalidateUserCaches(userID int) {
	s.algorithmService.InvalidateUserAlgorithmCaches(userID)
}

// GetNearbyUsers returns users within a specified distance
func (s *MatchService) GetNearbyUsers(userID int, maxDistanceKm int, limit int) ([]MatchResult, error) {
	request := &MatchingRequest{
		UserID:      userID,
		Algorithm:   AlgorithmProximity,
		Limit:       limit,
		MaxDistance: &maxDistanceKm,
	}
	return s.algorithmService.RunMatchingAlgorithm(request)
}

// GetRandomMatches returns random compatible users
func (s *MatchService) GetRandomMatches(userID int, limit int) ([]MatchResult, error) {
	request := &MatchingRequest{
		UserID:    userID,
		Algorithm: AlgorithmRandom,
		Limit:     limit,
	}
	return s.algorithmService.RunMatchingAlgorithm(request)
}

// GetNewUsers returns recently joined users
func (s *MatchService) GetNewUsers(userID int, limit int, daysBack int) ([]MatchResult, error) {
	request := &MatchingRequest{
		UserID:   userID,
		Algorithm: AlgorithmNewUsers,
		Limit:    limit,
		DaysBack: &daysBack,
	}
	return s.algorithmService.RunMatchingAlgorithm(request)
}

// GetPopularUsers returns users with high fame ratings
func (s *MatchService) GetPopularUsers(userID int, limit int, minFame int) ([]MatchResult, error) {
	request := &MatchingRequest{
		UserID:   userID,
		Algorithm: AlgorithmPopular,
		Limit:    limit,
		MinFame:  &minFame,
	}
	return s.algorithmService.RunMatchingAlgorithm(request)
}

// ValidateUser checks if a user exists
func (s *MatchService) ValidateUser(userID int) error {
	return s.userService.ValidateUserExists(userID)
}

// ReceivedLike represents a like received by a user
type ReceivedLike struct {
	ID           int         `json:"id"`
	UserID       int         `json:"user_id"`
	TargetUserID int         `json:"target_user_id"`
	UserProfile  MatchResult `json:"user_profile"`
	CreatedAt    string      `json:"created_at"`
	IsMutual     bool        `json:"is_mutual"`
}

// ReceivedLikePreview represents a blurred preview for free users
type ReceivedLikePreview struct {
	ID                string `json:"id"`
	CreatedAt         string `json:"created_at"`
	BlurredImage      string `json:"blurred_image"`
	TimestampRelative string `json:"timestamp_relative"`
}

// LikeStats represents statistics about received likes
type LikeStats struct {
	TotalLikesReceived   int     `json:"total_likes_received"`
	LikesToday          int     `json:"likes_today"`
	LikesThisWeek       int     `json:"likes_this_week"`
	LikesThisMonth      int     `json:"likes_this_month"`
	MostLikedPhoto      *string `json:"most_liked_photo,omitempty"`
	LikeRateTrend       string  `json:"like_rate_trend"`
	AverageLikesPerDay  float64 `json:"average_likes_per_day"`
}

// GetReceivedLikes returns all likes received by a user (Premium feature)
func (s *MatchService) GetReceivedLikes(userID int) ([]ReceivedLike, error) {
	return s.interactionService.GetReceivedLikes(userID)
}

// GetReceivedLikesPreview returns limited preview of received likes
func (s *MatchService) GetReceivedLikesPreview(userID int, limit int) ([]ReceivedLikePreview, error) {
	return s.interactionService.GetReceivedLikesPreview(userID, limit)
}

// GetLikeStats returns statistics about likes received by the user
func (s *MatchService) GetLikeStats(userID int) (*LikeStats, error) {
	return s.interactionService.GetLikeStats(userID)
}